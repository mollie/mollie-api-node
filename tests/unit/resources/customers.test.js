import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import Customers from 'resources/customers';
import Customer from 'models/customer';

import response from '../__stubs__/customers.json';

const mock = new MockAdapter(axios);

const props = {
  id: 'cst_kEn1PlbGa',
  name: 'Customer A',
  email: 'customer@example.org',
};

describe('customers', () => {
  let customers;
  beforeEach(() => {
    customers = new Customers(axios.create());
  });

  it('should have a resource name and model', () => {
    expect(Customers.resource).toBe('customers');
    expect(Customers.model).toBe(Customer);
  });

  describe('.create()', () => {
    mock.onPost('/customers').reply(200, response._embedded.customers[0]);

    it('should return a customer instance', () =>
      customers.create(props).then(result => {
        expect(result).toBeInstanceOf(Customer);
        expect(result.name).toBe(props.name);
        expect(result.email).toBe(props.email);
      }));

    it('should work with a callback', done => {
      customers.create(props, (err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Customer);
        expect(result.name).toBe(props.name);
        expect(result.email).toBe(props.email);
        done();
      });
    });
  });

  describe('.get()', () => {
    const error = { error: { message: 'The customer id is invalid' } };

    mock.onGet(`/customers/${props.id}`).reply(200, response._embedded.customers[0]);
    mock.onGet('/customers/foo').reply(500, error);

    it('should return a customer instance', () =>
      customers.get(props.id).then(result => {
        expect(result).toBeInstanceOf(Customer);
        expect(result.id).toBe(props.id);
      }));

    it('should work with a callback', done => {
      customers.get(props.id, (err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Customer);
        expect(result.id).toBe(props.id);
        done();
      });
    });

    it('should return an error for non-existing IDs', () =>
      customers
        .get('foo')
        .then(() => {
          throw new Error('Should reject');
        })
        .catch(err => {
          expect(err).toBe(error);
        }));

    it('should return an error with a callback for non-existing IDs', done => {
      customers.get('foo', (err, result) => {
        expect(err).toBe(error);
        expect(result).toBeUndefined();
        done();
      });
    });
  });

  describe('.update()', () => {
    const error = { error: { message: 'The customer id is invalid' } };

    mock.onPost(`/customers/${props.id}`).reply(200, response._embedded.customers[0]);
    mock.onPost('/customers/foo').reply(500, error);

    it('should return a customer instance', () =>
      customers.update(props.id, props).then(result => {
        expect(result).toBeInstanceOf(Customer);
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      customers.update(props.id, props, (err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Customer);
        expect(result).toMatchSnapshot();
        done();
      });
    });

    it('should return an error for non-existing IDs', () =>
      customers
        .update('foo')
        .then(() => {
          throw new Error('Should reject');
        })
        .catch(err => {
          expect(err).toBe(error);
        }));

    it('should return an error with a callback for non-existing IDs', done => {
      customers.update('foo', (err, result) => {
        expect(err).toBe(error);
        expect(result).toBeUndefined();
        done();
      });
    });
  });

  describe('.all()', () => {
    mock.onGet('/customers').reply(200, response);

    it('should return a list of all customers', () =>
      customers.all().then(result => {
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      customers.all((err, result) => {
        expect(err).toBeNull();
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
        done();
      });
    });
  });
});
