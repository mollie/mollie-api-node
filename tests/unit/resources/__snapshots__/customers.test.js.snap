// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customers .all() should return a list of all customers 1`] = `
Array [
  Customer {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-04-06T13:23:21.0Z",
    "email": "customer@example.org",
    "id": "cst_kEn1PlbGa",
    "locale": "nl_NL",
    "metadata": null,
    "mode": "test",
    "name": "Customer A",
    "recentlyUsedMethods": Array [
      "creditcard",
      "ideal",
    ],
    "resource": "customer",
  },
]
`;

exports[`customers .all() should work with a callback 1`] = `
Array [
  Customer {
    "_links": Object {
      "documentation": Object {
        "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
        "type": "text/html",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
        "type": "application/hal+json",
      },
    },
    "createdAt": "2018-04-06T13:23:21.0Z",
    "email": "customer@example.org",
    "id": "cst_kEn1PlbGa",
    "locale": "nl_NL",
    "metadata": null,
    "mode": "test",
    "name": "Customer A",
    "recentlyUsedMethods": Array [
      "creditcard",
      "ideal",
    ],
    "resource": "customer",
  },
]
`;

exports[`customers .update() should return a customer instance 1`] = `
Customer {
  "_links": Object {
    "documentation": Object {
      "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
      "type": "text/html",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
      "type": "application/hal+json",
    },
  },
  "createdAt": "2018-04-06T13:23:21.0Z",
  "email": "customer@example.org",
  "id": "cst_kEn1PlbGa",
  "locale": "nl_NL",
  "metadata": null,
  "mode": "test",
  "name": "Customer A",
  "recentlyUsedMethods": Array [
    "creditcard",
    "ideal",
  ],
  "resource": "customer",
}
`;

exports[`customers .update() should work with a callback 1`] = `
Customer {
  "_links": Object {
    "documentation": Object {
      "href": "https://docs.mollie.com/reference/v2/customers-api/get-customer",
      "type": "text/html",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_kEn1PlbGa",
      "type": "application/hal+json",
    },
  },
  "createdAt": "2018-04-06T13:23:21.0Z",
  "email": "customer@example.org",
  "id": "cst_kEn1PlbGa",
  "locale": "nl_NL",
  "metadata": null,
  "mode": "test",
  "name": "Customer A",
  "recentlyUsedMethods": Array [
    "creditcard",
    "ideal",
  ],
  "resource": "customer",
}
`;
